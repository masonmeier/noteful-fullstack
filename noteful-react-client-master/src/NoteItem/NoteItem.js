import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';

import config from '../config';
import NotefulContext from '../NotefulContext';

// this function 1st deletes via the API, then from state
// context.deleteNote = the update function, to update state in context
// Using a CLASS instead of a functional component I do *** NOT *** have to
//  pass in PROPS to get history, location and match (AND props.note.id) AND pass in CONTEXT for deleteNote function because there is no other way for the handleClickDelete function
// - to know what is in Context
// - to know history, location and match

export default class NoteItem extends React.Component {
	// need to grab NotefulContext (globals)
	static contextType = NotefulContext;

	handleClickDelete = e => {
		e.preventDefault();

		const noteId = this.props.note.id;

		fetch(config.NOTES_ENDPOINT + `/${noteId}`, {
			method: 'DELETE',
			headers: {
				'content-type': 'application/json',
				authorization: `Bearer ${config.API_KEY}`
			}
		})
			.then(res => {
				// I think b/c cors, typecode gives a res.status = 404 and an EMPTY error object when try to delete note so,
				if (!res.ok || res.status === '404') {
					// get the error message from the response,
					return res.json().then(error => {
						// then throw it
						// throw res.status instead of error b/c error is an empty object
						throw res.status;
					});
				}
				return res.json();
			})
			.then(data => {
				// call the callback function when the request is successful
				// this is where the App component can remove it from state
				// ie. update the notes stored in state
				// which also updates the notes stored in context
				this.context.deleteNote(noteId);

				// if in Note detail, return to show all notes list
				if (this.props.location.pathname.includes('/notes/')) {
					this.props.history.push(`/`);
				}

				// if have selected a folder, need to refresh after delete a note
				if (this.props.location.pathname.includes('/folders/')) {
					this.props.history.push(this.props.location.pathname);
				}
			})
			.catch(error => {
				// WORKAROUND to handle EMPTY error object and res.status = 404
				if (error !== 404) {
					this.context.addErrorNotes(error);
				}

				if (error === 404) {
					this.context.deleteNote(noteId);

					// if in Note detail, return to show all notes list
					if (this.props.location.pathname.includes('/notes/')) {
						this.props.history.push(`/`);
					}

					// if have selected a folder, need to refresh after delete a note
					if (this.props.location.pathname.includes('/folders/')) {
						this.props.history.push(this.props.location.pathname);
					}
				}
			});
	};

	render() {
		// to see noteItem notesError in ui & trigger NoteError boundary:
		// this.context.addErrorNotes({ value: 'notesItemAPI errorMessage' });
		return (
			<div className="note-item">
				{this.context.notesError && (
					<p className="error">{this.context.notesError.value}</p>
				)}
				{/*
			THIS CAUSED A staticContent ERROR:
			<NavLink to={`/notes/${note.id}`} {...props}>
				<h3>{note.title}</h3>
			</NavLink> */}
				<NavLink
					to={{
						pathname: `/notes/${this.props.note.id}`,
						props: this.props
					}}
				>
					<h3>{this.props.note.name}</h3>
				</NavLink>
				<div className="button-container">
					<span>
						Modified on{' '}
						<span className="note-datemod">
							{' '}
							{this.props.note.modified
								? this.props.note.modified.toString().split('T', 1)[0]
								: ''}
						</span>
					</span>
					<span>
						<NavLink to={`/edit-note/${this.props.note.id}`}>
							<button className="btn-edit-note">&#9998;</button>
						</NavLink>
						&nbsp;&nbsp;&nbsp;
						<button
							className="btn-delete-note"
							onClick={this.handleClickDelete}
						>
							-
						</button>
					</span>
				</div>
			</div>
		);
	}
}

// to catch bugs
// check that get a notes array that has id, name, and modified
// this array is the "notes" variable coming from context
NoteItem.propTypes = {
	notes: PropTypes.arrayOf(
		PropTypes.shape({
			id: PropTypes.string.isRequired,
			name: PropTypes.string.isRequired,
			modified: PropTypes.instanceOf(Date).isRequired
		})
	)
};

// default values
// ONLY for props that are NOT required
// so do NOT create NoteItem.defaultProps
// NoteItem.defaultProps = {
// 	note: {
// 		id: '',
// 		name: '',
// 		modified: ''
// 	}
// };
