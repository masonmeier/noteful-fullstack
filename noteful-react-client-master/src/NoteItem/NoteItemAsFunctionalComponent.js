import React, { useContext } from 'react';
import { NavLink } from 'react-router-dom';

import config from '../config';
import NotefulContext from '../NotefulContext';
import {createApiUrl} from '../utils/api';

// this function 1st deletes via the API, then from state
// context.deleteNote = the update function, to update state in context
// must pass in PROPS to get history, location and match (AND props.note.id) AND pass in CONTEXT for deleteNote function because there is no other way for the handleClickDelete function
// - to know what is in Context
// - to know history, location and match

function handleClickDelete(props, context) {
	const noteId = props.note.id;

	fetch(createApiUrl(`/notes/${noteId}`), {
		method: 'DELETE',
		headers: {
			'content-type': 'application/json'
		}
	})
		.then(res => {
			if (!res.ok) {
				// get the error message from the response,
				return res.json().then(error => {
					// then throw it
					throw error;
				});
			}
			return res.json();
		})
		.then(data => {
			// call the callback function when the request is successful
			// this is where the App component can remove it from state
			// ie. update the notes stored in state
			// which also updates the notes stored in context
			context.deleteNote(noteId);

			// if in Note detail, return to show all notes list
			if (props.location.pathname.includes('/notes/')) {
				props.history.push(`/`);
			}
		})
		.catch(error => {
			context.addErrorNotes(error);
		});
}

const NoteItem = props => {
	// need to grab NotefulContext (globals)
	const contextType = useContext(NotefulContext);

	return (
		<NotefulContext.Consumer>
			{/*
			Use the Consumer to grab values from contex
			--- the value we're grabbing from context is the deleteNote function, we're passing it to the delete button

 			QUESTION: what is context?
			ANSWER:
			is it equal to the object inside NotefulCcontext.js?
			an anonymous function with the parameter context automatically defined by {context => ... }, don't know where the parameter function comes from, don't need to know where it comes from because we're using it to render a result; context is a variable containing data; this function is called by line 39 NotefulContext.Consumer

			NOTE: context could be any word since its just the parameter label
			*/}

			{context => (
				<div className="note-item">
					{/*
			THIS CAUSED A staticContent ERROR:
			<NavLink to={`/notes/${note.id}`} {...props}>
				<h3>{note.title}</h3>
			</NavLink> */}

					<NavLink to={{ pathname: `/notes/${props.note.id}`, props: props }}>
						<h3>{props.note.name}</h3>
					</NavLink>

					<div className="button-container">
						<span>
							Modified on{' '}
							<span className="note-datemod">
								{props.note.modified
									? props.note.modified.split('T', 1)[0]
									: ''}
							</span>
						</span>
						<span>
							<button
								className="btn-delete"
								onClick={() => {
									handleClickDelete(props, contextType);
								}}
							>
								-
							</button>
						</span>
					</div>
				</div>
			)}
		</NotefulContext.Consumer>
	);
};
export default NoteItem;
