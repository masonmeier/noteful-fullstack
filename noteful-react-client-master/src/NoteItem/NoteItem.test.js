import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import NoteItem from './NoteItem';

describe(`NoteItem component`, () => {
	const props = {
		note: {
			id: 1,
			id_folder: 101,
			title: 'test-note-title',
			modified: new Date(2019, 8, 30)
		}
	};

	it('renders a NoteItem by default', () => {
		const wrapper = shallow(<NoteItem {...props} />);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('renders the NoteItem given props', () => {
		const wrapper = shallow(<NoteItem {...props} />);
		expect(toJson(wrapper)).toMatchSnapshot();
	});
});
