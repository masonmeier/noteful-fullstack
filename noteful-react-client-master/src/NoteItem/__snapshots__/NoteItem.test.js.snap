// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteItem component renders a NoteItem by default 1`] = `
<div
  className="note-item"
>
  <NavLink
    to={
      Object {
        "pathname": "/notes/1",
        "props": Object {
          "note": Object {
            "id": 1,
            "id_folder": 101,
            "modified": 2019-09-30T04:00:00.000Z,
            "title": "test-note-title",
          },
        },
      }
    }
  >
    <h3 />
  </NavLink>
  <div
    className="button-container"
  >
    <span>
      Modified on
       
      <span
        className="note-datemod"
      >
         
        Mon Sep 30 2019 00:00:00 GM
      </span>
    </span>
    <span>
      <NavLink
        to="/edit-note/1"
      >
        <button
          className="btn-edit-note"
        >
          ✎
        </button>
      </NavLink>
         
      <button
        className="btn-delete-note"
        onClick={[Function]}
      >
        -
      </button>
    </span>
  </div>
</div>
`;

exports[`NoteItem component renders the NoteItem given props 1`] = `
<div
  className="note-item"
>
  <NavLink
    to={
      Object {
        "pathname": "/notes/1",
        "props": Object {
          "note": Object {
            "id": 1,
            "id_folder": 101,
            "modified": 2019-09-30T04:00:00.000Z,
            "title": "test-note-title",
          },
        },
      }
    }
  >
    <h3 />
  </NavLink>
  <div
    className="button-container"
  >
    <span>
      Modified on
       
      <span
        className="note-datemod"
      >
         
        Mon Sep 30 2019 00:00:00 GM
      </span>
    </span>
    <span>
      <NavLink
        to="/edit-note/1"
      >
        <button
          className="btn-edit-note"
        >
          ✎
        </button>
      </NavLink>
         
      <button
        className="btn-delete-note"
        onClick={[Function]}
      >
        -
      </button>
    </span>
  </div>
</div>
`;
